router/batchrouter/asyncdestinationmanager/salesforce-bulk/salesforce_bulk.go

```diff
@@ -0,0 +1,460 @@
+	"github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+)
+
+func (s *SalesforceBulkUploader) Transform(job *jobsdb.JobT) (string, error) {
+	// Parse the event payload directly
+	if !gjson.ValidBytes(job.EventPayload) {
+		return "", fmt.Errorf("invalid JSON in event payload")
+	}
+
+	// Extract required fields from the input
+	traits := gjson.GetBytes(job.EventPayload, "traits")
+	externalId := gjson.GetBytes(job.EventPayload, "context.externalId")
+
+	// Build the metadata object
+	metadata := make(map[string]interface{})
+	metadata["job_id"] = float64(job.JobID)
+
+	// We are supporting only upsert operation
+	metadata["rudderOperation"] = "upsert"
+
+	// Add externalId to metadata if it exists
+	var externalIdArray []interface{}
+	if externalId.Exists() {
+		if err := jsonrs.Unmarshal([]byte(externalId.Raw), &externalIdArray); err == nil {
+			metadata["externalId"] = externalIdArray
+		}
+	}
+
+	// Build the message object
+	message := make(map[string]interface{})
+
+	// Add all traits to message
+	if traits.Exists() && traits.IsObject() {
+		traits.ForEach(func(key, value gjson.Result) bool {
+			message[key.String()] = value.Value()
+			return true
+		})
+	}
+
+	externalIdObject, err := extractFromVDM(externalIdArray)
+	if err != nil {
+		return "", fmt.Errorf("failed to extract externalId: %w", err)
+	}
+	// Add externalId field to message
+	message[externalIdObject.ExternalIDField] = externalIdObject.ExternalIDValue
+
+	// Create the final AsyncJob structure
+	asyncJob := common.AsyncJob{
+		Message:  message,
+		Metadata: metadata,
+	}
+
+	// Marshal and return
+	responsePayload, err := jsonrs.Marshal(asyncJob)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal async job: %w", err)
+	}
+
+	return string(responsePayload), nil
+}
+
+func (s *SalesforceBulkUploader) Upload(asyncDestStruct *common.AsyncDestinationStruct) common.AsyncUploadOutput {
+	destination := asyncDestStruct.Destination
+	destinationID := destination.ID
+	filePath := asyncDestStruct.FileName
+	failedJobIDs := asyncDestStruct.FailedJobIDs
+	importingJobIDs := asyncDestStruct.ImportingJobIDs
+
+	input, err := readJobsFromFile(filePath)
+	if err != nil {
+		return common.AsyncUploadOutput{
+			FailedJobIDs:  append(failedJobIDs, importingJobIDs...),
+			FailedReason:  fmt.Sprintf("Error reading jobs from file: %v", err),
+			FailedCount:   len(failedJobIDs) + len(importingJobIDs),
+			DestinationID: destinationID,
+		}
+	}
+
+	objectInfo, err := extractObjectInfo(input)
+	if err != nil {
+		return common.AsyncUploadOutput{
+			FailedJobIDs:  append(failedJobIDs, importingJobIDs...),
+			FailedReason:  fmt.Sprintf("Error extracting object info: %v", err),
+			FailedCount:   len(failedJobIDs) + len(importingJobIDs),
+			DestinationID: destinationID,
+		}
+	}
+
+	var allImportingJobIDs []int64
+	var allFailedJobIDs []int64
+	var sfJobs []SalesforceJobInfo
+
+	allFailedJobIDs = append(allFailedJobIDs, failedJobIDs...)
```

Comment:
Why are we appending failedJobIDs in here? At this points asyncDestStruct.FailedJobIDs will be empty right?